WMF Framework/CacheDBWriterHelper.swift

```diff
@@ -0,0 +1,118 @@
+
+import Foundation
+
+final class CacheDBWriterHelper {
+    static func fetchOrCreateCacheGroup(with groupKey: String, in moc: NSManagedObjectContext) -> PersistentCacheGroup? {
+        return cacheGroup(with: groupKey, in: moc) ?? createCacheGroup(with: groupKey, in: moc)
+    }
+
+    static func fetchOrCreateCacheItem(with itemKey: String, in moc: NSManagedObjectContext) -> PersistentCacheItem? {
+        return cacheItem(with: itemKey, in: moc) ?? createCacheItem(with: itemKey, in: moc)
+    }
+    
+    static func inMemoryCacheGroup(with key: String, in moc: NSManagedObjectContext) -> PersistentCacheGroup? {
+        for object in moc.registeredObjects where !object.isFault {
+            let predicate = NSPredicate(format: "key == %@", key)
+            guard let result = object as? PersistentCacheGroup, predicate.evaluate(with: result) else {
+                continue
+            }
+            return result
+        }
+        return nil
+    }
+
+    static func cacheGroup(with key: String, in moc: NSManagedObjectContext) -> PersistentCacheGroup? {
+        
+        if let inMemoryGroup = inMemoryCacheGroup(with: key, in: moc) {
+            return inMemoryGroup
+        }
+        
+        let fetchRequest: NSFetchRequest<PersistentCacheGroup> = PersistentCacheGroup.fetchRequest()
+        fetchRequest.predicate = NSPredicate(format: "key == %@", key)
+        fetchRequest.fetchLimit = 1
+        do {
+            guard let group = try moc.fetch(fetchRequest).first else {
+                return nil
+            }
+            return group
+        } catch let error {
+            fatalError(error.localizedDescription)
+        }
+    }
+    
+    static func createCacheGroup(with groupKey: String, in moc: NSManagedObjectContext) -> PersistentCacheGroup? {
+        
+        guard let entity = NSEntityDescription.entity(forEntityName: "PersistentCacheGroup", in: moc) else {
+            return nil
+        }
+        let group = PersistentCacheGroup(entity: entity, insertInto: moc)
+        group.key = groupKey
+        return group
+    }
+    
+    static func cacheItem(with itemKey: String, in moc: NSManagedObjectContext) -> PersistentCacheItem? {
+        
+        if let inMemoryItem = inMemoryCacheItem(with: itemKey, in: moc) {
+            return inMemoryItem
+        }
+        
+        let fetchRequest: NSFetchRequest<PersistentCacheItem> = PersistentCacheItem.fetchRequest()
+        fetchRequest.predicate = NSPredicate(format: "key == %@", itemKey)
+        fetchRequest.fetchLimit = 1
+        do {
+            guard let item = try moc.fetch(fetchRequest).first else {
+                return nil
+            }
+            return item
+        } catch let error {
+            fatalError(error.localizedDescription)
+        }
+    }
+    
+    static func inMemoryCacheItem(with key: String, in moc: NSManagedObjectContext) -> PersistentCacheItem? {
```

Comment:
Is there an advantage to this over just running the fetch?

Reply:
@joewalsh just an optimization I had read about, this could potentially speed things up by avoiding the fetch request and thus the hit to the database if the object is likely already in memory. One change I made recently is try to limit references to PersistentCacheItem & PersistentCacheGroup only to the dbWriters, I thought the separation was nice but that means I'm passing only the keys around and potentially fetching twice in the same flow.  So one flow being:

1. dbWriter creates (or fetches if they already exist) CacheItems
2. fileWriter to fetches & save files from CacheItem keys 
3. dbWriter fetches CacheItems from keys and marks isDownloaded=true

Step 3 could avoid hitting the database because those CacheItems are hopefully already in memory from step 1. I didn't performance test to confirm it's needed though, just thought of in the moment.

Reply:
@tonisevener I added some logging to measure the time - it looks like after enough `NSManagedObjects` are in memory (after saving) it's about the same speed as the fetch and both options are quick enough. I'd say it's OK to remove the optimization for now and we can revisit (maybe with a constant time lookup) if performance becomes a concern

Reply:
👍 sounds good

Reply:
This has been removed now.